includes/environment.inc

```diff
@@ -323,6 +323,11 @@ function drush_locate_root($start_path = NULL) {
     if ($follow_symlinks && is_link($path)) {
       $path = realpath($path);
     }
+    // Check to see if we have any bootstrap classes in this location.
+    $bootstrap_class_dir = $path . '/drush/bootstrap';
+    if (is_dir($bootstrap_class_dir)) {
+      _drush_add_commandfiles(array($bootstrap_class_dir), DRUSH_BOOTSTRAP_DRUSH);
```

Comment:
Loading the bootstrap object is an integral part of finding the root; we couldn't really move it somewhere else. The best we could do is to refactor drush_locate_root into two functions, and make the first one drush_locate_bootstrap_class.  This would have a performance impact, since we'd be walking up the tree twice instead of once. Perhaps this would not be significant due to filesystem caching.  I can do the refactor if you think it significant.

Also, I could refactor _drush_add_commandfiles so that we had another entrypoint into it that searched for a more limited pattern of commandfiles, e.g. foo.drush-bootstrap.inc, or whatever.  This would avoid accidentally loading general-purpose commandfiles that were mis-placed in this directory.  We could also improve performance by limiting recursive searching at this point -- which would only be a problem if a user made the mistake of misplacing a large folder in a 'bootstrap' directory.